include/nbl/video/alloc/SubAllocatedDescriptorSet.h

```diff
@@ -23,26 +23,66 @@ class SubAllocatedDescriptorSet : public core::IReferenceCounted
 	using value_type = typename AddressAllocator::size_type;
 	static constexpr value_type invalid_value = AddressAllocator::invalid_address;
 
+	class DeferredFreeFunctor
+	{
+	public:
+		inline DeferredFreeFunctor(SubAllocatedDescriptorSet* composed, uint32_t binding, size_type count, value_type* addresses)
+			: m_addresses(addresses, addresses + count), m_binding(binding), m_composed(composed)
+		{
+		}
+
+		// Just does the de-allocation
+		inline void operator()()
+		{
+			// isn't assert already debug-only?
+			#ifdef _NBL_DEBUG
+			assert(m_composed);
+			#endif // _NBL_DEBUG
+			m_composed->multi_deallocate(m_binding, m_addresses.size(), &m_addresses[0]);
+		}
+
+		// Takes count of allocations we want to free up as reference, true is returned if
+		// the amount of allocations freed was >= allocationsToFreeUp
+		// False is returned if there are more allocations to free up
+		inline bool operator()(size_type allocationsToFreeUp)
+		{
+			auto prevCount = m_addresses.size();
+			operator()();
+			auto totalFreed = m_addresses.size() - prevCount;
+
+			// This does the same logic as bool operator()(size_type&) on 
+			// CAsyncSingleBufferSubAllocator
+			return totalFreed >= allocationsToFreeUp;
+		}
+	protected:
+		SubAllocatedDescriptorSet* m_composed;
+		uint32_t m_binding;
+		std::vector<value_type> m_addresses;
+	};
 protected:
 	struct SubAllocDescriptorSetRange {
 		std::shared_ptr<AddressAllocator> addressAllocator;
 		std::shared_ptr<ReservedAllocator> reservedAllocator;
 		size_t reservedSize;
 	};
+	MultiTimelineEventHandlerST<DeferredFreeFunctor> eventHandler;
 	std::map<uint32_t, SubAllocDescriptorSetRange> m_allocatableRanges = {};
+	core::smart_refctd_ptr<video::IGPUDescriptorSet> m_descriptorSet;
 
 	#ifdef _NBL_DEBUG
 	std::recursive_mutex stAccessVerfier;
 	#endif // _NBL_DEBUG
 
-	constexpr static inline uint32_t MaxDescriptorSetAllocationAlignment = 64u*1024u; // if you need larger alignments then you're not right in the head
+	constexpr static inline uint32_t MaxDescriptorSetAllocationAlignment = 1u; 
 	constexpr static inline uint32_t MinDescriptorSetAllocationSize = 1u;
 
 public:
+
 	// constructors
 	template<typename... Args>
-	inline SubAllocatedDescriptorSet(video::IGPUDescriptorSetLayout* layout)
+	inline SubAllocatedDescriptorSet(video::IGPUDescriptorSet* descriptorSet)
```

Comment:
always take single pointers you want to take shared ownership of as `core::smart_refctd_ptr<T>&&` into functions like this

Also remove the `template<typename... Args>`